From 2db0284a3b104d653e7afa810aa3c03e23a0047f Mon Sep 17 00:00:00 2001
From: Patrick Bellasi <patrick.bellasi@arm.com>
Date: Mon, 22 Jun 2015 18:32:36 +0100
Subject: [PATCH 16/38] WIP: sched/fair: add function to convert boost value
 into "margin"

The basic idea of the boost knob is to "artificially inflate" some signals
to make a Task/RQ appears more demanding than what actually it is.
Independently from the specific signal, a consistent and possibly simple
semantic for the concept of "signal boosting" should be defined.
Such a semantic must defined:
1. how we translate the boost percentage into a "margin" value to be added
   to the original signal to be inflated
2. what is the meaning of a boost value from a user-space perspective

This patch provides the implementation of a possible boost semantic,
named Signal Proportional Compensation (SPC), where the boost percentage
(BP) is used to compute a margin (M) which is proportional to the
complement of the original signal (OS):
  M = BP * (1024 - OS)
The computed margin then added to the OS to obtain the Boosted Signal (BS)
  BS = OS + M

The proposed boost semantic has these main features:
- each signals gets a boost which is proportional to its delta with respect
  to the maximum available capacity in the system
- a 100% boosting has a clear understanding from a user-space perspective,
  since it means simply to run (possibly) "all" tasks at the max OPP
- each boosting value means to improve the tasks performance by a quantity
  which is proportional to the maximum achievable performance on that
  system.
Thus this semantics is somehow forcing a behaviour which is:

  50% boosting means to run at half-way between the current and the
    maximum performance which a task could achieve on that system

This patch provides the code to implement a fast integer division to convert
a boost percentage (BP) value into a margin (M).

NOTE: this code is suitable for all signals operating in range
      [0..SCHED_LOAD_SCALE]

Change-Id: I0e5bf863e7844300ef3744cff02d97502669949a
Signed-off-by: Patrick Bellasi <patrick.bellasi@arm.com>
---
 kernel/sched/fair.c | 39 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 71c0181..e3cf6b5 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -5231,6 +5231,45 @@ static inline unsigned long task_utilization(struct task_struct *p)
 	return p->se.avg.utilization_avg_contrib;
 }
 
+#ifdef CONFIG_SCHED_TUNE
+
+static unsigned long
+schedtune_margin(unsigned long signal, unsigned long boost)
+{
+	unsigned long long margin = 0;
+
+	/*
+	 * Signal proportional compensation (SPC)
+	 *
+	 * The Boost (B) value is used to compute a Maring (M) which is
+	 * proportional to the complement of the original Signal (S):
+	 *   M = B * (1024-S)
+	 * The obtained M could be used by the caller to "boost" S.
+	 */
+	margin  = SCHED_LOAD_SCALE - signal;
+	margin *= boost;
+
+	/*
+	 * Fast integer division by constant:
+	 *  Constant   :                 (C) = 100
+	 *  Precision  : 0.1%            (P) = 0.1
+	 *  Reference  : C * 100 / P     (R) = 100000
+	 *
+	 * Thus:
+	 *  Shift bifs : ceil(log(R,2))  (S) = 17
+	 *  Mult const : round(2^S/C)    (M) = 1311
+	 *
+	 *
+	 * */
+	margin  *= 1311;
+	margin >>= 17;
+
+	return margin;
+
+}
+
+#endif /* CONFIG_SCHED_TUNE */
+
 static inline bool __task_fits(struct task_struct *p, int cpu, int usage)
 {
 	unsigned long capacity = capacity_of(cpu);
-- 
1.9.1

